Return derivation path when answering `isOurs`

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#2176 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- 6c3aaead33829e9b3b87d012fd50ccfc21d19aa9
  :round_pushpin: **change derivation path JSON serialization to be less verbose**
    And also aligned with other interfaces like cardano-addresses.

- 1bba794e45c40249da3483ee732a5fa6be260b03
  :round_pushpin: **change 'isOurs' to return a derivation path instead of a boolean**
    This can then be used to figure out what are the derivation path of
  a bunch of addresses when returning raw coin-selections.
  Note that this commit builds but is so-to-speak unsound. We need to
  find a way to feed the purpose, coin type and account index down to
  the 'isOurs' function. The most logical place to do this is as part of
  the state. We can't use arbitrary constant here because both Icarus
  and Shelley use a SeqState, but have different purpose indexes.

- 4b116de1d5a1f8d3c4f33bc12b6eaca1774a6deb
  :round_pushpin: **store seq-state derivation prefix in the database.**
    That prefixes tells us which account corresponds to which state and also, which purpose so can distinguish between Icarus and Shelley wallets. This will require a database migration
  which I'll add in a later commit.

- 185d8fb4ccf8c395614c7f8dd6ffcb1c34b09476
  :round_pushpin: **define manual migrations for seq-state with regards to the derivation prefix**
    I've generated databases for Icarus and Shelley wallets from the latest master and, in a test now trying to open these database and observe that a) it is possible, b) there's a log line indicating that a migration has happened, c) the resulting prefix in each database is exactly what we expect it to be.


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
change derivation path JSON serialization to be less verbose

  And also aligned with other interfaces like cardano-addresses.

KtorZ committed:
change 'isOurs' to return a derivation path instead of a boolean

  This can then be used to figure out what are the derivation path of
  a bunch of addresses when returning raw coin-selections.
  Note that this commit builds but is so-to-speak unsound. We need to
  find a way to feed the purpose, coin type and account index down to
  the 'isOurs' function. The most logical place to do this is as part of
  the state. We can't use arbitrary constant here because both Icarus
  and Shelley use a SeqState, but have different purpose indexes.

KtorZ committed:
store seq-state derivation prefix in the database.

  That prefixes tells us which account corresponds to which state and also, which purpose so can distinguish between Icarus and Shelley wallets. This will require a database migration
  which I'll add in a later commit.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Jormungandr.hs:
```diff
@@ -436,8 +410,17 @@ instance MkKeyFingerprint JormungandrKey (Proxy (n :: NetworkDiscriminant), Jorm
+            purpose :: Index 'Hardened 'PurposeK
+            coinType :: Index 'Hardened 'CoinTypeK
+            accountIx :: Index 'Hardened 'AccountK
+            (purpose, coinType, accountIx) = undefined -- TODO: have this in the seq state
```
Whoops. Forgot to change that one. Will do.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:
```diff
@@ -350,8 +322,17 @@ instance MkKeyFingerprint ShelleyKey (Proxy (n :: NetworkDiscriminant), ShelleyK
+            purpose :: Index 'Hardened 'PurposeK
+            coinType :: Index 'Hardened 'CoinTypeK
+            accountIx :: Index 'Hardened 'AccountK
+            (purpose, coinType, accountIx) = undefined -- TODO: have this in the seq state
```
Whoops. Forgot to change that one. Will do.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -551,6 +555,8 @@ data SeqState (n :: NetworkDiscriminant) k = SeqState
     , rewardAccountKey :: k 'AddressK XPub
         -- ^ Reward account public key associated with this wallet
+    , derivationPrefix :: DerivationPrefix
+        -- ^ Derivation path prefix from a root key up to the internal account
```
There are two major changes in this PR (which then ripples to many other files). One of them is here. `DerivationPrefix` is defined below and is simply a 3-tuple of the first three derivation indexes that leads to the account associated with that state. 

This is necessary because we use seq-state for both Shelley and Icarus which have different `purpose` levels. Also, this will build a nice foundation for the up-coming multi-account support by having the actual account index within the state itself too!

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -632,17 +707,35 @@ instance
+                (_, Just addrIx) ->
+                    Just (purpose, coinType, accountIx, utxoInternal, addrIx)
+
+                _ -> Nothing
```
The other main change is here, in the definition of `isOurs`. Instead of returning a mere `Bool`, we now return the full derivation path corresponding to that address.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs:
```diff
@@ -202,14 +203,16 @@ instance RndStateLike (RndState n) where
+    type DerivationPath (RndState n) Address = DerivationPath
     isOurs addr st =
-        (isJust path, maybe id (addDiscoveredAddress addr Used) path st)
+        (path, maybe id (addDiscoveredAddress addr Used) path st)
```
For random addresses, this is trivially achieved because decrypting the path is exactly how we identify whether a random address is ours or not.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -568,13 +574,76 @@ instance
+                <*> fromText coinTypeT
+                <*> fromText accountT
+            _ ->
+                Left $ TextDecodingError "expected exactly 3 derivation paths"
```
TODO: add roundtrip tests to test these instances.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -227,11 +227,12 @@ UTxO                                sql=utxo
+    seqStateInternalGap       W.AddressPoolGap   sql=internal_gap
+    seqStateAccountXPub       B8.ByteString      sql=account_xpub
+    seqStateRewardXPub        B8.ByteString      sql=reward_xpub
+    seqStateDerivationPrefix  W.DerivationPrefix sql=derivation_prefix
```
TODO: This will require a database migration. Will add in a later commit.

KtorZ replied:
Done.

paweljakubas reviewed (approved):
Very fond of renaming initiative taken here - `purposeCIP1852`, `coinTypeAda`,.. all give more clarity. There is **some more work** to do (roundtrips, defined undefined, db migration support..etc) already mentioned. I wonder if some things (like `purposeCI1852`) should not be (some day in future) relocated further to cardano-addresses and imported here in cardano-wallet

KtorZ committed:
define manual migrations for seq-state with regards to the derivation prefix

  I've generated databases for Icarus and Shelley wallets from the latest master and, in a test now trying to open these database and observe that a) it is possible, b) there's a log line indicating that a migration has happened, c) the resulting prefix in each database is exactly what we expect it to be.

KtorZ committed:
add roundtrip test for ToText/FromText DerivationPrefix instances

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery.hs:
```diff
@@ -55,10 +55,11 @@ import Cardano.Wallet.Primitive.Types
         :: entity
         -> s
-        -> (Bool, s)
+        -> (Maybe (DerivationPath s entity), s)
```
Comment below could be more descriptive:

```
         -- ^ Returns derivation path if the entity is ours, otherwise Nothing.
```

hasufell commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -954,9 +960,40 @@ instance FromJSON ApiAddressDerivationPath where
 instance ToJSON ApiAddressDerivationSegment where
-    toJSON = genericToJSON defaultRecordTypeOptions
+    toJSON (ApiAddressDerivationSegment (ApiRelativeAddressIndex ix) typ)
+        | typ == Hardened = toJSON (show ix <> "H")
```
`show` makes me feel uneasy, since it's for debugging purposes mostly, not serialisation, although this use case seems safe.

There's a more elaborate way via Text: https://github.com/haskell/text/issues/218#issue-296457899
```hs
T.toStrict . B.toLazyText . B.decimal $ ix
```

KtorZ replied:
Yes, for Int / Word type, show is rather safe so I usually don't bother. The result is the same and quite more readable than more elaborated ways.

KtorZ replied:
I think we could also just go for `toText` here since we have such instances for Word / Int I believe. Adding them otherwise would be nice.

hasufell commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -954,9 +960,40 @@ instance FromJSON ApiAddressDerivationPath where
-    toJSON = genericToJSON defaultRecordTypeOptions
+    toJSON (ApiAddressDerivationSegment (ApiRelativeAddressIndex ix) typ)
+        | typ == Hardened = toJSON (show ix <> "H")
+        | otherwise = toJSON (show ix)
```
same as above

hasufell commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -227,11 +227,12 @@ UTxO                                sql=utxo
+    seqStateInternalGap       W.AddressPoolGap   sql=internal_gap
+    seqStateAccountXPub       B8.ByteString      sql=account_xpub
+    seqStateRewardXPub        B8.ByteString      sql=reward_xpub
+    seqStateDerivationPrefix  W.DerivationPrefix sql=derivation_prefix
```
Maybe I'm being pedantic, but do we need to align all these rows? It increases the diff.

KtorZ replied:
We don't need to. It helps readability, but is arguably superfluous.

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -148,7 +151,7 @@ import qualified Data.Text.Encoding as T
 -- We do not manipulate purpose, cointype and change paths directly, so they are
 -- left out of the sum type.
-data Depth = RootK | AccountK | AddressK
+data Depth = RootK | PurposeK | CoinTypeK | AccountK | RoleK | AddressK
```
It seems the documentation doesn't apply anymore.

KtorZ replied:
Indeed!

hasufell commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -568,13 +574,76 @@ instance
       where
         chgsF = blockListF' "-" build (pendingIxsToList chgs)
 
+newtype DerivationPrefix = DerivationPrefix
```
Maybe it's obvious what this is to others, but some documentation could still help?

KtorZ replied:
Definitely.

KtorZ replied:
Added:

```hs
-- | Each 'SeqState' is like a bucket of addresses associated with an 'account'.
-- An 'account' corresponds to a subset of an HD tree as defined in BIP-0039.
--
-- cardano-wallet implements two similar HD schemes on top of BIP-0039 that are:
--
-- - BIP-0044 (for so-called Icarus wallets)
-- - CIP-1815 (for so-called Shelley and Jormungandr wallets)
--
-- Both scheme works by considering 5 levels of derivation from an initial root
-- key (see also 'Depth' from Cardano.Wallet.Primitive.AddressDerivation). A
-- SeqState keeps track of indexes from the two last levels of a derivation
-- branch. The 'DerivationPrefix' defines the first three indexes chosen for
-- this particular 'SeqState'.
```

KtorZ committed:
adjust swagger documentation to reflect the presence derivation path in resolved inputs.

KtorZ committed:
change isOurs to return an homogenous list instead of n-tuples of indexes.

hasufell commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -369,21 +375,21 @@ data ApiAddress (n :: NetworkDiscriminant) = ApiAddress
-data ApiAddressDerivationSegment = ApiAddressDerivationSegment
-    { derivationIndex :: !ApiRelativeAddressIndex
-    , derivationType :: !ApiAddressDerivationType
+data ApiDerivationSegment = ApiDerivationSegment
```
I want to use this data type in the `Cardano.Wallet`, which must not use API types. Because this is similar to 

```hs
newtype Index (derivationType :: DerivationType) (level :: Depth) = Index
    { getIndex :: Word32 }
```
which I can't use as in a homogenous List easily. `ApiAddressDerivationSegment` would basically be non-typelevel version of `Index`.

hasufell replied:
Or... we could define a toText instance for Index

KtorZ replied:
What about using the `DerivationIndex` from the `Primitive.Types` instead ?

hasufell replied:
That doesn't preserve the derivation type, don't we need that?

KtorZ replied:
When considering all indexes as a list (or path):

- the depth matters not because what the `Depth` captures is actually the position of the index in that list. 
- the derivationType is captured by representing indexes as plain Word32. The `Soft` / `Hardened` notation is for easing human-readability but in the end, a soft index is simply a value < 2^31, whereas a "hardened" index is simply a value >= 2^31. Therefore, instead of representing indexes as derivationType + relative index within 0 and 2^31, we can represent them as just an index between 0 and 2^32, which is what `DerivationIndex` does.

KtorZ replied:
I should probably add these bits as comments.

KtorZ committed:
add '*.faulty.json' to gitignore, which we should've done a while ago already

KtorZ committed:
cleanup now redundant API types regarding derivation path

  Turns out that it is much easier and less error-prone to use a single
  opaque type 'DerivationIndex' to represent derivation path as homogeneous
  lists. Having 2 level of indirections in the API now makes thing more
  verbose and quite hard to grasp for a non-educated reader.

KtorZ committed:
extend integration scenario about coin-selection with some assertions about the derivation path.
